packages/ts-transformers/src/internal/decorators/query-assigned-elements.ts

```
@@ -0,0 +1,255 @@
+
+  constructor({factory}: ts.TransformationContext) {
+    this._factory = factory;
+  }
+
+  visit(
+    litClassContext: LitClassContext,
+    property: ts.ClassElement,
+    decorator: ts.Decorator
+  ) {
+    if (!ts.isPropertyDeclaration(property)) {
+      return;
+    }
+    if (!ts.isCallExpression(decorator.expression)) {
+      return;
+    }
+    if (!ts.isIdentifier(property.name)) {
+      return;
+    }
+    const name = property.name.text;
+    const [arg0] = decorator.expression.arguments;
+
+    if (arg0 && !ts.isObjectLiteralExpression(arg0)) {
+      throw new Error(
+        `queryAssignedElements argument is expected to be an inlined ` +
+          `object literal. Instead received: '${arg0.getText()}'`
+      );
+    }
+
+    if (arg0 && arg0.properties.some((p) => !ts.isPropertyAssignment(p))) {
+      throw new Error(
+        `queryAssignedElements object literal argument can only include ` +
+          `property assignment. For example: '{ slotName: "example" }' is ` +
+          `supported, whilst '{ ...otherOpts }' is unsupported.`
+      );
+    }
+
+    const {slotName, selector} = this._retrieveSlotAndSelector(arg0);
+    litClassContext.litFileContext.replaceAndMoveComments(
+      property,
+      this._createQueryAssignedElementsGetter(
+        name,
+        slotName,
+        selector,
+        this._filterAssignedElementsOptions(arg0)
+      )
+    );
+  }
+
+  private _retrieveSlotAndSelector(opts?: ts.ObjectLiteralExpression): {
+    slotName: string;
+    selector: string;
+  } {
+    if (!opts) {
+      return {slotName: '', selector: ''};
+    }
+    const findStringLiteralFor = (key: string): string => {
+      const propAssignment = opts.properties.find(
+        (p) => p.name && ts.isIdentifier(p.name) && p.name.text === key
+      );
+      if (
+        propAssignment &&
+        ts.isPropertyAssignment(propAssignment) &&
+        ts.isStringLiteral(propAssignment.initializer)
```

Should probably throw if it's not a string literal, because it could also be a variable and we'd silently ignore that here.

---

Done thank you!

---

Actually just occured to me that `selector` could be used as-is whatever kind of node it is, since it just goes directly into the `filter` arguments, right?

And you could maybe also handle `slot` being an identifier or whatever too by generating `` `slot[name=${node}]` `` in the case that it's not a string?